### ISSUE TITLE
[Bert/PyTorch] Question about SyncFreeStats and Loss Gather

### ISSUE THREAD

**Issue by Hannibal046** [NVIDIA/DeepLearningExamples#1119]
Hi, thanks for your great work of reproduction of Bert Pretraining. After reading the source code of `run_pretraining.py`. I am confused about a few things. It would be so nice if you guys can take time to help me out. My questions are:
1. What does SyncFreeStates do ? I think this is a class to collect necessary information when training like `loss`,`lr`,`steps` etc. Am I right ? But why there are `host` and `device` in the class ? Do you have any special design here ?
2. Where does the loss get averaged when logging ? Is it because **Sync**FreeStates ? But I can't tell it .
3. Since you already register a hook to do gradient accumulation division, why still divide when logging:
https://github.com/NVIDIA/DeepLearningExamples/blob/3e8897f7855ff475d69699e66c8f668f6191dfa4/PyTorch/LanguageModeling/BERT/run_pretraining.py#L714
Thanks again for your help !

**Comment by sharathts** [comment]
1. yes, that is correct. It is a class used to gather training stats in a sync-free way on each device/GPU. (Hence host and device in each class). Each device creates a class object to collect stats.

2. During training, we log loss/stats from device:0 only. https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/run_pretraining.py#L712
   After all steps of training, we log the averages loss across all devices.(we don't average with intermediate logging to simply prevent an allreduce for logging)
   
 3. hooks divides the gradients. The loss is still undivided. Hence the division before logging.

**Comment by Hannibal046** [comment]
Hi, thanks for response. Just to be clear, this only logs main process loss which is just an approximation of real loss in consideration of communication cost. Am I right ?
https://github.com/NVIDIA/DeepLearningExamples/blob/3e8897f7855ff475d69699e66c8f668f6191dfa4/PyTorch/LanguageModeling/BERT/run_pretraining.py#L711-L716

**Comment by Hannibal046** [comment]
BTW, there is no gradient clipping in bert pretraining, am I correct ?

**Comment by sharathts** [comment]
Yes, main process loss is an approximation of real loss. 
But the final loss that is logged after the training run is the real(averaged across all devices) loss.

We clip gradients to a max norm of 1.0 in the LAMB optimizer : https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/run_pretraining.py#L418, https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/lamb_amp_opt/fused_lamb/fused_lamb.py#L17

**Comment by Hannibal046** [comment]
Thanks so much. It solved all my problems
